fdbclient/ManagementAPI.actor.cpp

```diff
@@ -3750,13 +3670,68 @@ ACTOR Future<Void> prepareExclusiveRangeLockOperation(Transaction* tr,
 			ASSERT(rangeLockStateSet.isValid());
 			auto lockSet = rangeLockStateSet.getLocks();
-			if (!lockSet.empty() && lockSet.find(newLock.getLockUniqueString()) == lockSet.end()) {
+			if (!lockSet.empty() && (!rangeLockStateSet.isLockedFor(RangeLockType::ExclusiveReadLock) ||
+			                         lockSet.find(RangeLockState(RangeLockType::ExclusiveReadLock, ownerUniqueID, range)
+			                                          .getLockUniqueString()) == lockSet.end())) {
 				TraceEvent(SevDebug, "PrepareExclusiveRangeLockOperationFailed")
 				    .detail("Reason", "Locked")
-				    .detail("NewLock", newLock.toString())
-				    .detail("ExistingLocks", rangeLockStateSet.toString())
-				    .detail("Range", range);
-				throw range_locked_by_different_user(); // Has been locked by a different owner
+				    .detail("NewLockType", RangeLockType::ExclusiveReadLock)
+				    .detail("NewLockRange", range)
+				    .detail("NewLockOwner", ownerUniqueID)
+				    .detail("ExistingLocks", rangeLockStateSet.toString());
+				throw range_lock_reject(); // Has been locked
+			}
+		}
+		beginKey = res[res.size() - 1].key;
+	}
+	return Void();
+}
+
+ACTOR Future<Void> prepareExclusiveRangeUnlockOperation(Transaction* tr,
+                                                        KeyRange range,
+                                                        RangeLockOwnerName ownerUniqueID) {
+	// Check input range
+	if (range.end > normalKeys.end) {
+		TraceEvent(SevDebug, "PrepareExclusiveRangeUnlockOperationFailed")
+		    .detail("Reason", "Range out of scope")
+		    .detail("Range", range);
+		throw range_lock_failed();
+	}
+	// Check owner
+	state Optional<Value> ownerValue = wait(tr->get(rangeLockOwnerKeyFor(ownerUniqueID)));
+	if (!ownerValue.present()) {
+		TraceEvent(SevDebug, "PrepareExclusiveRangeUnlockOperationFailed")
+		    .detail("Reason", "Owner not found")
+		    .detail("Owner", ownerUniqueID)
+		    .detail("Range", range);
+		throw range_lock_failed();
+	}
+	state RangeLockOwner owner = decodeRangeLockOwner(ownerValue.get());
+	ASSERT(owner.isValid());
+
+	// Check lock state on the entire input range. Throw exception if the range has been locked by a different owner.
+	state Key beginKey = range.begin;
+	state Key endKey = range.end;
+	state KeyRange rangeToRead;
+	while (beginKey < endKey) {
+		rangeToRead = KeyRangeRef(beginKey, endKey);
+		RangeResult res = wait(krmGetRanges(tr, rangeLockPrefix, rangeToRead));
+		for (int i = 0; i < res.size() - 1; i++) {
+			if (res[i].value.empty()) {
+				continue;
+			}
+			RangeLockStateSet rangeLockStateSet = decodeRangeLockStateSet(res[i].value);
+			ASSERT(rangeLockStateSet.isValid());
+			auto lockSet = rangeLockStateSet.getLocks();
+			if (!lockSet.empty() && (!rangeLockStateSet.isLockedFor(RangeLockType::ExclusiveReadLock) ||
+			                         lockSet.find(RangeLockState(RangeLockType::ExclusiveReadLock, ownerUniqueID, range)
```

Comment:
is this condition needed? `lockSet.empty()` is true here?

Reply:
Yes, it is needed. When `!rangeLockStateSet.isLockedFor(RangeLockType::ExclusiveReadLock) ||  lockSet.find(RangeLockState(RangeLockType::ExclusiveReadLock, ownerUniqueID, range).getLockUniqueString()) == lockSet.end()` is true, it is possible that the lockSet is empty. However, when lockSet empty, there is no lock on the range. In this case, we do not want to reject request of locking/releasing the lock.